crates/viewer/re_ui/src/filter_widget.rs

```diff
@@ -0,0 +1,276 @@
+                            // we add additional spacing for aesthetic reasons (active text edits have a
+                            // fat border)
+                            ui.spacing_mut().text_edit_width =
+                                (ui.max_rect().width() - 10.0).at_least(0.0);
+
+                            let response =
+                                egui::TextEdit::singleline(&mut inner_state.filter_query)
+                                    .lock_focus(true)
+                                    .ui(ui);
+
+                            if self.request_focus {
+                                self.request_focus = false;
+                                response.request_focus();
+                            }
+                        } else {
+                            ui.label(galley);
+                        }
+                    })
+                    .with_content_width(text_width)
+                    .action_button(icon, || {
+                        button_clicked = true;
+                    }),
+                );
+        });
+
+        // defer button handling because we can't mutably borrow `self` in both closures above
+        if button_clicked {
+            if self.inner_state.is_none() {
+                self.inner_state = Some(InnerState::default());
+                self.request_focus = true;
+            } else {
+                self.inner_state = None;
+            }
+        }
+    }
+}
+
+// --
+
+/// Full-text, case-insensitive matcher.
+pub struct FilterMatcher {
+    /// The lowercase version of the query string.
+    ///
+    /// If this is `None` or `Some("")`, the matcher will accept any input.
+    lowercase_query: Option<String>,
+}
+
+impl FilterMatcher {
+    fn new(query: Option<&str>) -> Self {
+        Self {
+            lowercase_query: query.map(|s| s.to_lowercase()),
+        }
+    }
+
+    /// Is the filter set to match everything?
+    ///
+    /// Can be used by client code to short-circuit more expansive matching logic.
+    pub fn matches_everything(&self) -> bool {
+        self.lowercase_query.is_none()
+    }
+
+    /// Is the filter set to match nothing?
+    ///
+    /// Can be used by client code to short-circuit more expansive matching logic.
+    pub fn matches_nothing(&self) -> bool {
+        self.lowercase_query
+            .as_ref()
+            .is_some_and(|query| query.is_empty())
+    }
+
+    /// Does the given text match the filter?
+    pub fn matches(&self, text: &str) -> bool {
+        match self.lowercase_query.as_deref() {
+            None => true,
+            Some("") => false,
+            Some(query) => text.to_lowercase().contains(query),
+        }
+    }
+
+    /// Find all matches in the given text.
+    ///
+    /// Can be used to format the text, e.g. with [`format_matching_text`].
+    pub fn find_matches(&self, text: &str) -> Option<impl Iterator<Item = (usize, usize)> + '_> {
```

Comment:
```suggestion
    pub fn find_matches(&self, text: &str) -> Option<impl Iterator<Item = Range<usize>)> + '_> {
```

Would make the semantic **a lot** clearer